Added GSDParser

rcortini opened the pull request:
This is a very preliminary attempt to add a parser for HOOMD-blue's new "GSD" trajectory/topology file format. This commit only includes a rough sketch of what the parser could look like, but currently I was not able to test it because I can't figure out how to have the package add this parser to the _PARSERS dictionary.

See my post on the mailing list
https://groups.google.com/forum/#!topic/mdnalysis-discussion/Ex8bbKWY6bM

richardjgowers commented:
Hi @rcortini 

If you subclass from `MDAnalysis.topology.base.TopologyReaderBase` then it should get automatically added to `_PARSERS`

richardjgowers commented on package/MDAnalysis/topology/GSDParser.py:
```diff
@@ -0,0 +1,135 @@
+"""
+from __future__ import absolute_import
+
+import gsd.hoomd
```
The tests are currently failing because you've added this as a dependency.  For now, you can wrap it in a try/except block to make things pass.  We'll look at how easy this is to install to decide if we just make this a requirement for MDA or not.

```python
try:
    import gsd.hoomd
except ImportError:
    pass
```

richardjgowers commented:
@rcortini give me a poke on here when it's ready for review.  It'll definitely need some tests for that though, here's an example of the minimal tests needed for coordinates:

https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/coordinates/test_txyz.py

rcortini commented:
@richardjgowers Thanks so much! I now have a working version of the trajectory and topology parser, and wrote a testsuite that hopefully looks like the one that it should be.

Please note that the GSD format was created with the possibility of varying topology in mind... now these parsers I wrote work for a system with fixed topology and number of particles, and no check is done to ensure that this is the case.

orbeckst commented:
> Please note that the GSD format was created with the possibility of varying topology in mind... now these parsers I wrote work for a system with fixed topology and number of particles, and no check is done to ensure that this is the case.

It's ok to start out with limited functionality but please
1. document the limitations
2. have some checks in place so that users don't get cryptic error messages

For the second part, it would be good to have a test case with a file that it _cannot_ process at the moment. The test should test that it fails in a graceful and predictable way.

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,133 @@
+    def _frame_to_ts(self, frame):
+        """convert a gsd-frame to a :class:`TimeStep`"""
+        ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+        ts.frame = frame.configuration.step
```
`ts.frame` **must** be `self._frame` (ie MDAnalysis' opinion of what the frame numbering is), if GSD also has an opinion on this, save it as `ts.data['step']` like here:

https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/XTC.py#L125

rcortini replied:
GSD stores the timestep of the simulation in `frame.configuration.step`, but I wasn't able to figure out a way of extracting the requested frame number in order to access the correct frame in the trajectory

richardjgowers replied:
So when MDAnalysis asks for `frame=2` it just wants the 3rd (because 0 based) record of positions that is stored in the file.  It's irrelevant how many timesteps have happened in the simulation, it might be the 3,000th timestep if you're saving every 1,000.  From what I can see, your `self._file[self._frame]` call looks correct (assuming GSD works as an iterable of frames too)

rcortini replied:
Okay now I set corrected this as you suggested. However, I'm getting as a result that if I slice the trajectory then the `ts.frame` is always zero. This was the reason for writing `ts.frame = frame.configuration.step` because I wasn't able to figure out a way *from inside* the `_frame_to_ts` function which number of frame was requested.

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,133 @@
+        """
+        return self.ts.dimensions
+
+    def __getitem__ (self,frame_numbers) :
```
You shouldn't have to override this, is there a reason the base version isn't working?

rcortini replied:
I did this because I got an error when attempting to access a trajectory slice. What's the correct way of doing it?

richardjgowers replied:
For a Reader you should only have to implement `_read_frame(frame)` and `_read_next_timestep()`. The base class takes care of handling `__getitem__` etc [here](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/coordinates/base.py#L1280), it takes care of stuff like boolean slicing etc and makes all the Readers behave the same way.

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,133 @@
+
+    def _frame_to_ts(self, frame):
+        """convert a gsd-frame to a :class:`TimeStep`"""
+        ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
```
this should reuse the old `self.ts` object

rcortini replied:
I'm sorry I got this wrong. I couldn't figure out the correct logic for this. In the DCD parser I see that the `__init__` method instantiates a Timestep object. Then the `_read_next_timestep` has a `ts = None` as argument, with a check on whether it is really `None` inside the code. Then the `_frame_to_ts` function has the Timestep object as an argument. However probably I don't understand how this works, since the `_frame_to_ts` function is called whenever one tries to access a frame information, and I don't see where the old `self.ts` object passed to that function in that case

richardjgowers commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -399,5 +400,7 @@
 ALIGN_BOUND = resource_filename(__name__, 'data/analysis/align_bound.pdb.gz')
 ALIGN_UNBOUND = resource_filename(__name__, 'data/analysis/align_unbound.pdb.gz')
 
+GSD = resource_filename(__name__, 'data/example.gsd')
```
the test file is 15.6MB, can this be trimmed to a smaller system/fewer frames

rcortini replied:
Sure, fixed

richardjgowers commented:
@rcortini sorry, I missed this before, but you're missing a method for a Reader.  They should have `_read_next_timestep` and `_read_frame(frame)`.  The first method just gets the next frame, the second jumps to an arbitrary frame and reads that.

If you do something like this it should work, where the method you've already written (`_frame_to_ts`) should be mostly unchanged; accepts a GSD frame and empties that into `self.ts`

```python
def _read_next_timestep(self):
    # be lazy and just reuse the other way :)
    return self._read_frame(self._frame + 1)

def _read_frame_(self, frame):
    # rather than checking if the frame *should* exist, just go for it
    try:
        myframe = self._file[frame]
    except IndexError:
        # and clean up if something goes wrong
        raise IOError
    self._frame_to_ts(myframe)
    return self.ts
    
```

rcortini commented:
@richardjgowers Okay I think I fixed the issues that you were mentioning, please have a look if you can

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    format = 'GSD'
+    units = {'time': None, 'length': None}
+    _Timestep = base.Timestep
```
this isn't needed

rcortini replied:
Okay, removed

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,123 @@
+        self.open_trajectory()
+        self.n_atoms = self._file[0].particles.N
+        self.ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+        self._read_next_timestep()
```
you're going to need to set `self.frame` here (to -1), else `_read_next` won't know what it is

rcortini replied:
Okay, fixed

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,123 @@
+
+    def open_trajectory (self) :
+        """opens the trajectory file using gsd.hoomd module"""
+        # self._frame = -1
```
this line is needed, else the Reader won't know what frame to start on

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,123 @@
+        .. versionadded:: 0.17.0
+        """
+        super(GSDReader, self).__init__(filename, **kwargs)
+        self.filename = filename
```
not needed, base class does this

rcortini replied:
I removed these lines and obtained the error
`AttributeError: 'GSDReader' object has no attribute 'filename'`

richardjgowers commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,123 @@
+        """read next frame in trajectory"""
+        return self._read_frame (self.frame + 1)
+
+    @property
```
We can't add dimensions here, it's a property of Timesteps/AtomGroups etc, but not Readers

rcortini replied:
Okay, removed

richardjgowers commented on testsuite/MDAnalysisTests/coordinates/test_gsd.py:
```diff
@@ -0,0 +1,38 @@
+def GSD_U():
+    return mda.Universe(GSD)
+
+def test_gsd_positions(GSD_U):
```
add some more tests for things like:
 - positions at the next frame
 - the total number of frames in the traj
 - the number of atoms in the file
 - dimensions
 - data['step']

rcortini replied:
I'll try to figure out how to do this based on other examples in the testsuite

richardjgowers commented on package/MDAnalysis/topology/GSDParser.py:
```diff
@@ -0,0 +1,138 @@
+)
+
+
+class GSDParser(TopologyReaderBase):
```
this class needs tests (separate to the reader tests).  check things like atomtypes, masses etc are all being read correctly.

rcortini replied:
you mean that these tests should be included in the testsuite, or inside the class?

richardjgowers commented:
@kain88-de this package is on a different conda channel (glotzer), is it worth making a recipe for conda-forge?

kain88-de commented:
> @kain88-de this package is on a different conda channel (glotzer), is it worth making a recipe for conda-forge?

This isn't to hard. They are very nice and have a working conda recipe in the repo. It can easily be adopted for conda-forge. Though I would prefer if the original authors did this themselves as they already seem to be actively publishing conda packages.

orbeckst commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,129 @@
+    def open_trajectory (self) :
+        """opens the trajectory file using gsd.hoomd module"""
+        self._frame = -1
+        self._file = gsd.hoomd.open (self.filename,mode='rb')
```
This fails if `gsd.hoomd` cannot be imported.


(Also: PEP8: no space between function name and parentheses)

orbeckst commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,129 @@
+try :
+    import gsd.hoomd
+except :
+    pass
```
`gsd.hoomd` appears to be required in `GSDReader`. 

At the moment the [import policy](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#list-of-core-module-dependencies) is that there are no optional dependencies in the core library so this has to be a hard import that fails if this package is not installed.

If we want to do this differently then we need to discuss it.

richardjgowers replied:
I think it's reasonable to have some formats not work out of the box if they have more difficult install routes (eg aren't on pypi).  We used to do this for netCDF before we started using scipy for that.

orbeckst commented on testsuite/CHANGELOG:
```diff
@@ -13,6 +13,11 @@ Also see https://github.com/MDAnalysis/mdanalysis/wiki/MDAnalysisTests
 ------------------------------------------------------------------------------
+11/02/17
+
+  * 0.17.0
```
Please rebase against develop. This looks as if you diffed against master.

orbeckst commented:
@richardjgowers @kain88-de did you look at the `gsd` package? Do we want it as a core dependency?

rcortini commented:
@orbeckst @richardjgowers I added the new tests, fixed the CHANGELOGs (I think I did a mistake in adding v 0.17.0 in the CHANGELOG of the testsuite, I fixed that), added the "Resnames" attribute to the GSDParser, and fixed the space between function names and parentheses.

richardjgowers commented:
@rcortini this is looking good, we're just sorting out the packaging of the gsd module.  Thanks for contributing!

kain88-de commented:
The gsd package is not available on pypi. This makes it not suitable as a core dependency. 

So I wouldn't like to have this in core. Having the GSD readers in MDAnalysis would still be nice anyway. It would be beneficial if we had a place in the code where we can place readers for formats that need additional dependencies or aren't as well supported as other formats.

richardjgowers commented:
Isn't ncdf already an optional dependency reader?

On Sat, 11 Nov 2017, 3:06 p.m. Max Linke, <notifications@github.com> wrote:

> The gsd package is not available on pypi. This makes it not suitable as a
> core dependency.
>
> So I wouldn't like to have this in core. Having the GSD readers in
> MDAnalysis would still be nice anyway. It would be beneficial if we had a
> place in the code where we can place readers for formats that need
> additional dependencies or aren't as well supported as other formats.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1693#issuecomment-343671286>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jB4TtI_w2UlA340Dq-q8Mj6M-ueImks5s1bfwgaJpZM4QMbwa>
> .
>

kain88-de commented:
ncdf works out of box since we use scipy for it. scipy is a core dependency.

orbeckst commented:
Could someone ask the Glotzer group to put the GSD package up on Pypi and condo-forge?

Oliver

> On Nov 11, 2017, at 8:06 AM, Max Linke <notifications@github.com> wrote:
> 
> The gsd package is not available on pypi. This makes it not suitable as a core dependency.
> 
> So I wouldn't like to have this in core. Having the GSD readers in MDAnalysis would still be nice anyway. It would be beneficial if we had a place in the code where we can place readers for formats that need additional dependencies or aren't as well supported as other formats.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/MDAnalysis/mdanalysis/pull/1693#issuecomment-343671286>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAOhnB--sLlgZtwp-k9bCeuunSqsL1uIks5s1bfvgaJpZM4QMbwa>.
> 

--
Oliver Beckstein
orbeckst@gmail.com * orbeckst@gmx.net

richardjgowers commented:
I asked about forge today

On Sat, 11 Nov 2017, 5:51 p.m. Oliver Beckstein, <notifications@github.com>
wrote:

> Could someone ask the Glotzer group to put the GSD package up on Pypi and
> condo-forge?
>
> Oliver
>
> > On Nov 11, 2017, at 8:06 AM, Max Linke <notifications@github.com> wrote:
> >
> > The gsd package is not available on pypi. This makes it not suitable as
> a core dependency.
> >
> > So I wouldn't like to have this in core. Having the GSD readers in
> MDAnalysis would still be nice anyway. It would be beneficial if we had a
> place in the code where we can place readers for formats that need
> additional dependencies or aren't as well supported as other formats.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub <
> https://github.com/MDAnalysis/mdanalysis/pull/1693#issuecomment-343671286>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AAOhnB--sLlgZtwp-k9bCeuunSqsL1uIks5s1bfvgaJpZM4QMbwa
> >.
> >
>
> --
> Oliver Beckstein
> orbeckst@gmail.com * orbeckst@gmx.net
>
>
>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1693#issuecomment-343681900>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jB_-KmqxTt4-V8lDXSQfrX7hkAW2Hks5s1d6NgaJpZM4QMbwa>
> .
>

joaander commented:
gsd is now available on conda-forge. Thank you for supporting HOOMD in MDAnalysis.

orbeckst commented:
@joaander did you also put up `gsd`  on PyPi so that `pip install` works? We are supporting both pip and conda with MDAnalysis.

kain88-de commented on package/MDAnalysis/coordinates/GSD.py:
```diff
@@ -0,0 +1,129 @@
+    def open_trajectory(self) :
+        """opens the trajectory file using gsd.hoomd module"""
+        self._frame = -1
+        self._file = gsd.hoomd.open(self.filename,mode='rb')
```
here it would be nice to have a good error message for the user. In general, in the init function you should make a test if the gsd module was available and then raise an error if it wasn't found. That is better then the error message this line will throw "no module gsd"

joaander commented:
@orbeckst I tried and failed to upload a gsd package to pypi. I don't have any further time to troubleshoot the failures at this point.

orbeckst commented:
@joaander sorry to hear that you won't be having a pip package available. We'd be happy to help.

Do you have a repository and/or issue tracker where we can raise this as an issue? Even if you don't fix it anytime soon, it will allow us to create an error message for our users that will direct them to the right place if they find that their MDAnalysis GSDParser failed. 

Do you have detailed installation instructions for users?

(We are not requiring conda and we are very reluctant to bring in dependencies that can fail at import time. If we do this then we need to be able to tell the users very clearly what  they need to be doing in order to get things working.)

joaander commented:
Repository: https://bitbucket.org/glotzer/gsd
Issue tracker: https://bitbucket.org/glotzer/gsd/issues?status=new&status=open
Reference documentation: http://gsd.readthedocs.io/en/latest/

README.md in the repository includes installation instructions for conda and full compilation instructions for both users and developers.

I also do not require conda, and I only provide conda binaries to my users as an expedient way for users to quickly get up and running. Most of my users compile and install manually (or managed by some script) so that the builds can utilize the native MPI and CUDA libraries on the clusters where the tools will be used. 

I recognize that GSD does not have these limitations, and so I am willing to support other installation methods if users will find them useful. However, I only have very limited time that I can apply to this low priority project. I followed a pypi tutorial and came up with this: https://bitbucket.org/glotzer/gsd/branch/pypi-test#diff - however it does not produce a package that works. The package is missing several files that are needed to build the c module. Within the time I had available, I could not find any documentation that described how I could correct this package.

orbeckst commented:
Many thanks! I appreciate your quick and comprehensive response, @joaander 

I definitely understand the time constraints and that supporting open source software can be very difficult. We'll see what we can do, maybe we have the expertise to make it work and can contribute a PR. 

Do you mind if, for the time being, we ping you on this thread if questions come up? Or do you prefer we opened an issue in your tracker?

joaander commented:
Feel free to contact me either way. I don't have a strong preference.

kain88-de commented:
[issue on bitbucket](https://bitbucket.org/glotzer/gsd/issues/7/publish-on-pypi)

joaander commented:
Thanks to some tips from Max Linke, gsd 1.4.0 is now published on pypi.

orbeckst commented:
With `gsd` on pypi and conda-forge available, we should be able to get this PR into core. 

Can someone manually test the pip installation? 

Sidenote: We probably should have one Travis job to do a minimal conda 3 installation and then let pip do most of the package pulling, just to check. (Maybe a Travis cron job?)

richardjgowers commented:
Ok I've added gsd to the travis build and our dependencies, should be good to go if the tests pass

richardjgowers commented:
closer/reopened to kickstart travis

orbeckst commented:
@richardjgowers tests still throwing a hissy-fit.
